src/main_nep/nep4.cu

```diff
@@ -258,62 +244,45 @@ static __global__ void apply_gnn_message_passing(
 {
   int n1 = threadIdx.x + blockIdx.x * blockDim.x;
   if (n1 < N) {
+    int t1 = g_type[n1];
     int num_neighbors_of_n1 = g_NN[n1];
-    const int F = ann.dim; // dimension of q_out, for now dim_out = dim_in.
-    for (int nu = 0; nu < F; nu++) {
-      float q_out_nu = g_messages[n1 + nu * N]; // fc(r_ii) = 1
-      // TODO perhaps normalize weights? Compare Kipf, Welling et al. (2016)
+    for (int nu = 0; nu < ann.dim; nu++) {
+      float q_out_nu = g_messages[n1 + nu * N];
       for (int j = 0; j < num_neighbors_of_n1; ++j) {
         int index = n1 + N * j;
         int n2 = g_NL[index];
+        int t2 = g_type[n2];
         float r12[3] = {g_x12[index], g_y12[index], g_z12[index]};
         float d12 = sqrt(r12[0] * r12[0] + r12[1] * r12[1] + r12[2] * r12[2]);
         float fc12;
         find_fc(para.rc_angular, para.rcinv_angular, d12, fc12);
+        fc12 *= ann.d[t1 * para.num_types + t2];
         q_out_nu += fc12 * g_messages[n2 + nu * N];
       }
-      gnn_descriptors[n1 + nu * N] = tanh(q_out_nu);
+      gnn_descriptors[n1 + nu * N] = q_out_nu;
     }
   }
 }
 
-// Precompute messages q*theta for all descriptors
-static __global__ void apply_gnn_compute_messages(
+static __global__ void apply_ann(
   const int N,
   const NEP4::ANN ann,
-  const NEP4::GNN gnn,
   const float* __restrict__ g_q,
-  const int* g_NN,
-  const int* g_NL,
-  float* gnn_messages)
-{
-  int n1 = threadIdx.x + blockIdx.x * blockDim.x;
-  if (n1 < N) {
-    const int F = ann.dim; // dimension of q_out, for now dim_out = dim_in.
-    for (int nu = 0; nu < F; nu++) {
-      float q_theta_nu = 0.0f;
-      for (int gamma = 0; gamma < ann.dim; gamma++) {
-        q_theta_nu += g_q[n1 + gamma * N] * gnn.theta[gamma + ann.dim * nu];
-      }
-      gnn_messages[n1 + nu * N] = q_theta_nu;
-    }
-  }
-}
-
-static __global__ void apply_ann(
-  const int N, const NEP4::ANN ann, const float* __restrict__ g_q, float* g_pe, float* g_dU_dq)
+  const float* __restrict__ g_q_scaler,
+  float* g_pe,
+  float* g_dU_dq)
 {
   int n1 = threadIdx.x + blockIdx.x * blockDim.x;
   if (n1 < N) {
     float q[MAX_DIM] = {0.0f};
     for (int d = 0; d < ann.dim; ++d) {
-      q[d] = g_q[n1 + d * N];
+      q[d] = g_q[n1 + d * N] * g_q_scaler[d];
```

Comment:
Why did you re-enable the scaling of the descriptors? I thought it wasn't needed in NEP3, but maybe I misunderstood you.

Reply:
Because I started this branch from master, which still has the scaling code. I suggested you remove it because I think it would be hard to figure out how to deal with the scaling for the derivatives. 

Now for my implementation, the scaling is for the message-passing descriptor, not the NEP3 descriptor and the scaling of the direvatives is as simple as before. So I simply kept it.

Reply:
Alright makes sense :+1: